Refactor of parsing scalar functions annotations code

fiedukow opened the pull request:
The general goal of this changes is to extract code responsible for parsing annotations of scalar functions and remove its duplicates.
All function generated from annotations do use only one implementation of SqlScalarFunction (ParametricScalar).

It's also distinguishable now from the code point of view, that we have two possibilities of declaring classes as ScalarFunctions providers. One case is when we declare class with single function and multiple specializations and the other is when we declare class with multiple different functions.

martint commented:
@cberner, can you review?

cberner commented:
Yep. Will be a little bit though, as I'm reviewing several other PRs at the moment

fiedukow commented:
Rebased to master.

cberner commented:
Looks like Travis is failing. Could you fix those and assign back to me?

fiedukow commented:
Done. Also resolved conflicts with current master.
Back to you @cberner

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:
```diff
@@ -144,7 +146,13 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
+                return new ScalarFunctionImplementation(implementation.isNullable(), implementation.getNullableArguments(), methodHandleAndConstructor.getMethodHandle(), methodHandleAndConstructor.getConstructor(), isDeterministic());
+            }
+            else {
+                return new ScalarFunctionImplementation(implementation.isNullable(), implementation.getNullableArguments(), implementation.getMethodHandle(), isDeterministic());
```
Why would `methodHandleAndConstructor` be `null`? It doesn't seem like that should happen for an "exact" implementation

fiedukow replied:
Thats correct, fixed.

cberner replied:
I still see it here. Do you mean it's fixed in a future commit? If so, you should make the change in this commit instead. We try to have each commit be self contained, and it should be possible to merge just a few commits out of a PR and still have the code base be in a "good" state.

fiedukow replied:
adressed in first commit, conflicts fixed.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:
```diff
@@ -173,40 +181,82 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
+            return ImmutableList.of(createReflectionParametricScalar(mangleOperatorName(operatorAnnotation.value()), true, true, descriptionAnnotation, findPublicMethodsWithAnnotation(clazz, SqlType.class), clazz.getSimpleName(), findConstructors(clazz)));
+        }
+        checkArgument(false);
+        return ImmutableList.of();
```
This is not reachable. Instead change the `checkArgument` to `throw new ...` with an appropriate message

fiedukow replied:
This code does not exists anymore.

cberner replied:
Same comment. You should remove it from this commit

fiedukow replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:
```diff
@@ -227,13 +277,20 @@ public static SqlScalarFunction parseDefinition(Class<?> clazz)
+            checkArgument(exactImplementationsMap.size() == 1, "It is not allowed to use clases without generic signature.");
+            Map.Entry<Signature, Implementation> onlyImplementation = exactImplementationsMap.entrySet().iterator().next();
+            signature = onlyImplementation.getKey();
+            //return SqlScalarFunction.create(onlyImplementation.getKey(), description, hidden, implementation.getMethodHandle(), Optional.empty(), deterministic, implementation.isNullable(), implementation.getNullableArguments());
```
remove

fiedukow replied:
outdated.

fiedukow replied:
done in outdated code

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:
```diff
@@ -173,40 +181,82 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
+    public static List<SqlScalarFunction> parseDefinition(Method method)
+    {
+        ScalarFunction scalarFunction = method.getAnnotation(ScalarFunction.class);
+        ScalarOperator scalarOperator = method.getAnnotation(ScalarOperator.class);
```
Check that only one of these is present

fiedukow replied:
Is there any reason not to allow annotation that defines scalar function and operator?

cberner replied:
Only that there doesn't seem to be a reason to allow it. It's hard to imagine a function where you would want to have functionality that's available as an operator also available as a function, so it's most likely a programming error if you have both

fiedukow replied:
Ok, done in historical commits.

fiedukow replied:
I think it is kinda good example:
```
    @ScalarFunction("date")
    @ScalarOperator(CAST)
    @SqlType(StandardTypes.DATE)
    public static long castFromSlice(@SqlType(StandardTypes.VARCHAR) Slice value)
    {
        try {
            return parseDate(trim(value).toStringUtf8());
        }
        catch (IllegalArgumentException e) {
            throw new PrestoException(INVALID_CAST_ARGUMENT, "Value cannot be cast to date: " + value.toStringUtf8(), e);
        }
    }
```

https://github.com/prestodb/presto/commit/189f97

cberner replied:
Oh interesting, didn't realize we had those. Ok, ya this makes sense then

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:
```diff
@@ -173,40 +181,82 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
 
-        Map<Set<TypeParameter>, Constructor<?>> constructors = findConstructors(clazz);
+        for (Method method : methods) {
+            if (name.isEmpty()) {
```
This can never happen. You have a check that name is non-empty at the beginning of this method

fiedukow replied:
outdated.

cberner replied:
It should be removed here, instead of in a future commit

fiedukow replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:
```diff
@@ -227,13 +277,20 @@ public static SqlScalarFunction parseDefinition(Class<?> clazz)
+        Map<Signature, Implementation> exactImplementationsMap = exactImplementations.build();
+        if (signature == null) {
+            checkArgument(!exactImplementationsMap.isEmpty(), "Implementation of ScalarFunction %s must be parametric or exact implementation.", objectName);
+            checkArgument(exactImplementationsMap.size() == 1, "It is not allowed to use clases without generic signature.");
```
The error message here isn't clear

fiedukow replied:
outdated.

fiedukow replied:
done in historical commit

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:
```diff
@@ -597,7 +654,6 @@ public MethodHandleAndConstructor specialize(Signature boundSignature, BoundVari
             if (nullable) {
                 return Primitives.wrap(clazz);
             }
-            checkArgument(clazz != void.class);
```
Did you change the handling for this somewhere? I don't see any changes related to it

fiedukow replied:
What is the reasoning behind this check?
Why do we allow `Void.class` and don't allow `void.class`?
Why simple function for (I assume) wrapping type into it's nullable version if it is expected to be nullable is not agnostic about the type that it's taking? I think if we don't wan't `void` to be allowed, we should check that on higher level and for `Void` as well.

cberner replied:
Looking at the code it seems like we should disallow both of them. Having a `void` native container type combined with a non-nullable parameter/return value doesn't make sense, as `NULL` is that only value that can be represented in a `void`

fiedukow replied:
I moved this check to higher level function where I think it belongs.

fiedukow replied:
https://github.com/prestodb/presto/blob/c7bf63a2138c2e61f0edb307cefbf6b35fba25c2/presto-main/src/main/java/com/facebook/presto/type/UnknownOperators.java

Well... Back to only checking for `void.class`. I would rather not check that at all, but I think sticking with what was already in the code is a safer way.

fiedukow replied:
Oh...
And there is that for `void.class`:
https://github.com/prestodb/presto/blob/c7bf63a2138c2e61f0edb307cefbf6b35fba25c2/presto-main/src/main/java/com/facebook/presto/operator/scalar/FailureFunction.java
It wasn't failing before as the `void.class` check was done in a wrong place and didn't cover all flows.
I think this check was strictly broken just dead for cases it mattered for.

I removed that check once again. I hope you don't mind that.

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionListBuilder.java:
```diff
@@ -62,8 +62,6 @@
 import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
```
This commit should be squashed into the previous one. We always remove dead code in the commit that it died in

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:
```diff
@@ -15,134 +15,114 @@
 
```
The rename of `ScalarHeaderAndImplementations` should be done in the previous commit, since that class was introduced in this PR

fiedukow replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ParametricScalar.java:
```diff
@@ -32,8 +34,8 @@
 {
-    ScalarHeader details;
-    ScalarImplementations implementations;
+    private final ScalarHeader details;
```
`private final` should be added in the commit which introduced these fields

fiedukow replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementation.java:
```diff
@@ -1,3 +1,16 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
```
This should be added in the commit that introduced this file

fiedukow replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarFromAnnotationsParser.java:
```diff
@@ -0,0 +1,184 @@
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.Objects.requireNonNull;
+
+public class ScalarFromAnnotationsParser
```
I'd name this `ScalarFunctionParser` the fact that it's "from annotations" is an implementation detail that's probably not important

fiedukow replied:
How is that implementation detail?
User of the class must know what does the parser parse to get Scalar. In this particular case it parses annotations.

cberner replied:
Ya, it just seems like a detail that's not important, but I don't have a strong preference, so we can leave it if you think this is more clear. My thinking though was that if this was, say, an Antlr parser that parsed text, you wouldn't name it `ScalarFromUtf8Parser`

fiedukow replied:
I think it's an "obvious" case, when a parser parses text :-)
If you don't have strong preference I would stick with current name.

cberner commented:
Had some comments

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:
```diff
@@ -100,44 +99,74 @@
+        private final ScalarImplementationHeader header;
+        private final List<Method> methods;
+
+        public ScalarHeaderAndImplementations(ScalarImplementationHeader header, List<Method> methods) {
```
`{` should be on next line. Can you run the Maven build for each commit? I think there's a style checker for this

fiedukow replied:
done (earlier, just forgot to write)

fiedukow commented:
Comments applied.
Rebased to master.
I've build with maven on every commit - some fixes regarding style were applied in the process.

Back to you @cberner

cberner commented:
Can you check that the comments are applied in the right commits? I still see some of them un-addressed. Alternately you can address the comments in a "fixup" commit that can be squashed before merging

fiedukow commented:
It is impossible "fixup" outdated commit if the code you want to change doesn't exists without going back to this commit, putting "fixup" there and then applying all other commits (resolving conflicts ofc).
Personally I think it wasn't worth as lot time as it took to fix this history and I may have broken more then I fixed while rebasing this multiple times to merge in fixes earlier in history.
Anyway - it is done now. Hopefully I didn't missed anything nor placed in wrong commit.

I think only three of your comments left unaddressed, as I’m waiting for your reaction:
https://github.com/prestodb/presto/pull/5377#discussion_r68763104
https://github.com/prestodb/presto/pull/5377#discussion_r68759265
https://github.com/prestodb/presto/pull/5377#discussion_r68928196

cberner commented:
Ya, it can definitely be difficult to go back and fix-up commits. The reason we do it that way is to make sure we can always `git bisect` and not worry about broken intermediate commits, and also so that part of a PR can be merged without merging the whole thing, or part of it reverted without leaving the tree in a bad state

cberner commented on presto-spi/src/main/java/com/facebook/presto/spi/StandardErrorCode.java:
```diff
@@ -49,6 +49,7 @@
     INVALID_PROCEDURE_ARGUMENT(0x0000_001E),
     QUERY_REJECTED(0x0000_001F),
     AMBIGUOUS_FUNCTION_CALL(0x0000_0020),
+    FORBIDDEN_FUNCTION_ANNOTATIONS(0x0000_0021),
```
Remove this since it's unused

fiedukow replied:
done

cberner commented:
`Fix ReflectionParametricScalar to cover all scalar cases` looks good, except for a few minor comments

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:
```diff
@@ -194,6 +194,20 @@ else if (operatorAnnotation != null) {
         throw new PrestoException(FUNCTION_NOT_FOUND, format("The class %s does not define neither @ScalarFunction nor @ScalarOperator.", clazz.getSimpleName()));
     }
 
+    public static List<SqlScalarFunction> parseFunctionSetClass(Class<?> clazz)
```
I think `parseFunctionDefinitions` would be a better name

fiedukow replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java:
```diff
@@ -178,7 +178,7 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
     }
 
-    public static List<SqlScalarFunction> parseDefinition(Class<?> clazz)
+    public static List<SqlScalarFunction> parseFunctionDefinitionClass(Class<?> clazz)
```
I would just name this `parseFunctionDefinition` it's obvious from the argument that it parses a `Class`

fiedukow replied:
done

cberner commented:
`Separate semanticaly different scalar classes` looks good also, just one stylistic comment

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementationHeader.java:
```diff
@@ -0,0 +1,117 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class ScalarImplementationHeader
```
Why have two separate classes `ScalarImplementationHeader` and `ScalarHeader`? It looks like `ScalarHeader` is only used inside `ScalarImplementationHeader`, so I would just merge them into a single class

fiedukow replied:
`ScalarHeader` is used in `ParametricScalar` too.

To clairfy:
`ScalarImplementationHeader` is all data obtained from parsed code and annotations. It describe function that can have multiple specializations so it does not contains argument information as `Signature` does.
`ScalarHeader` is "metadata" of ScalarMethod that is obtained from annotations which is not part of `Signature` (in fact it's `ScalarImplementationHeader - name`).
`Signature` is what functions system requires but nothing more than that (includes `name`).

This is separated that way mostly to avoid duplicate of `name` inside `ParametricScalar` (both `ScalarImplementationHeader` and `Signature` do contain `name`). `Signature` is forced there by the rest of the functions system.
Naming is sketchy here especially the `ScalarImplementationHeader` seems not to be precise enough - I'm more than interested in listing to propositions of better names :-)

cberner commented:
`Move annotation classes to annotations package` looks good

cberner commented:
`Avoid using null in favor of Optional in ScalarImplementation.specialize` looks good

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarFromAnnotationsParser.java:
```diff
@@ -0,0 +1,184 @@
+
+        ScalarImplementations implementations = new ScalarImplementations(exactImplementations.build(), specializedImplementations.build(), genericImplementations.build());
+
+        return new ParametricScalar(
```
style note: you can put all the arguments on one line

fiedukow replied:
done.

cberner commented:
Looks good. Had a few minor comments

fiedukow commented:
I think all your comments except
https://github.com/prestodb/presto/pull/5377#discussion_r68763104
https://github.com/prestodb/presto/pull/5377#discussion_r68995025
are fixed now.

Back to you for acceptance/further discussion.

fiedukow commented:
Rebase to master is required as it fails builds right now. Back to me.

fiedukow commented:
New commit is added https://github.com/prestodb/presto/pull/5377/commits/fef2795ecf2e6088e9bc7c1467a403121318e3d4
It is possible to do that due to this changes together with (just integrated) varchar(x) support in aggregates.

It may be squashed with previous commit but I wanted to keep it separate to show what changed in more clear way - at least for the time of review.

I hope I didn't break anything, but as there were some conflicts it might have happened. I've run check if all commits do build and it seems like they do (I didn't run complete tests on all of them though). We should definitely wait for travis to pass before merging this.

Back to you.

fiedukow commented:
Still needs fixes after rebase.

fiedukow commented:
There was additional conflict with dfa16d9b that what introduced yesterday after my last rebase.
Well, it's fixed now, I hope the build will go smoothly now.

fiedukow commented:
@cberner 
I removed previous comments (starting at yours last one) and summarize current status here as it passes the Travis check right now and I think it's mergeable again.

I applied all your comments except of:
https://github.com/prestodb/presto/pull/5377#discussion_r68763104
https://github.com/prestodb/presto/pull/5377#discussion_r68995025
https://github.com/prestodb/presto/pull/5377#discussion_r68457161
Those three are more and less discussed but if you still disagree with any of those - please continue the discussion.

I rebased everything to master (twice) as the Travis check were failing due to hidden conflicts. I fixed problems that caused that.

This patches together with varchar(x) support in aggregation allowed to do one more refactor which is introduced in 09171b9. This is totally new commit in this PR. It may be squashed with previous one, but I left it as separate change to make it clear what has changed. I think this PR cannot be merged without this additional change, as it will leave `typeManager` variable unused.

I've checked that each commit still do build and passes style checks (`./mvnw install -T C1 -pl '!presto-docs,!presto-server-rpm' -DskipTests=true -Dmaven.javadoc.skip=true -q`)

As there were many minor changes, and new commit with major change was added, I've reluctantly removed `accepted` label, as I think it is right thing to do :(

Now it's all in yours hands again :-)
Please treat this PR with priority as there are other PR waiting for that to be merged - thanks in advance!

electrum commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:
```diff
@@ -364,44 +364,44 @@ public WindowFunctionSupplier load(SpecializedFunctionKey key)
                 .aggregate(CovarianceAggregation.class)
                 .aggregate(RegressionAggregation.class)
                 .aggregate(CorrelationAggregation.class)
-                .scalar(SequenceFunction.class)
```
Let's leave the name "scalar" to be consistent with the other methods. Also, to minimize unnecessary changes.

fiedukow replied:
I think it is good change:
 1) If you wan't this name to be the same than we have to restore code for recognizing which version do we wan't. This causes many tricky scenarios and lowers the possibility of finding user error while parsing this.
 2) This are two semantically different classes types. We don't ever use same class for both of this things. Code recognizing that should check for all different behaviors like defining some parts of both classes types. It's tricky to make it bullet proof. Of course we don't need it bullet proof. However by separating that (so that it matches the semantic), we have that for free - user defines what does he try to do and if he do something different checks will force him to fix this.
 3) Philosophically: I think one function doing two different things is against OOP rules.
 4) By the definition of the word, you may call every refactoring `unnecessary`, but I don't think there is value in minimizing refactors as such.
 5) To be consistent I would rather continue this changes to the rest of methods (I'm about to start refactoring + extending annotation support for aggregate functions as a followup of https://github.com/prestodb/presto/pull/5381

electrum replied:
I thought this was a pure rename. I didn't realize there were two different versions of the function.

cberner commented:
@electrum see @fiedukow 's comment. I don't have an opinion here (the new names seemed fine to me), so please discuss with him and then assign back to me.

electrum commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionListBuilder.java:
```diff
@@ -87,21 +83,13 @@ public FunctionListBuilder aggregate(Class<?> aggregationDefinition)
-            }
-        }
-        checkArgument(foundOne, "Expected class %s to be annotated with @%s, or contain at least one method annotated with @%s", clazz.getName(), ScalarFunction.class.getSimpleName(), ScalarFunction.class.getSimpleName());
+    public FunctionListBuilder scalars(Class<?> clazz)
```
This function is a duplicate of `scalar`, so it should be a delegate (if we have it at all).

cberner replied:
It's actually not. They call different functions (note the plural "definitions")

cberner commented:
Merged. Thanks!

fiedukow commented:
Thats great!
Thank you for both review and patience! ;-)
